### PR TITLE
QL: disallow comparison between fields in combination with arithmetic operations

### PR DESCRIPTION
**Pull request by astefan**
Both EQL and SQL  disallow fields comparison, even if those fields are found in functions. This PR fixes a scenario where the fields are used inside arithmetic operations. For example `length(first_name) + lenght(last_name) > 10` start being forbidden as a result of this fix (`length(first_name) > length(last_name)` was already forbidden).

### CODE REVIEW

**Review by costin — APPROVED** [review]
LGTM

**Review by matriv — APPROVED** [review]
LGTM, it's good to fix inconsistencies.

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/planner/ExpressionTranslators.java [hunk 1]
```diff
@@ -264,9 +267,44 @@ protected Query asQuery(BinaryComparison bc, TranslatorHandler handler) {
...
+            bc.left().forEachDown(ArithmeticOperation.class, op -> checkFieldsUsageInArithmeticOperation(op, bc));
+        }
+
+        @SuppressWarnings("rawtypes")
+        public static void checkFieldsUsageInArithmeticOperation(ArithmeticOperation a, BinaryPredicate bc) {
+            Set<FieldAttribute> leftFields = collectFields(a.left(), bc);
+            Set<FieldAttribute> rightFields = collectFields(a.right(), bc);
+            boolean containsAll = (leftFields.size() > rightFields.size() && leftFields.containsAll(rightFields))
+                || (leftFields.size() < rightFields.size() && rightFields.containsAll(leftFields))
+                || leftFields.containsAll(rightFields);
```
Comment (bpintea) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/planner/ExpressionTranslators.java:282]: Nit: couldn't the last condition be done away with by adding the equality to the the first condition?

Comment (astefan) [reply x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/planner/ExpressionTranslators.java:282]: Indeed, nice catch. Will change.

File: x-pack/plugin/sql/qa/server/src/main/resources/arithmetic.sql-spec [hunk 2]
```diff
@@ -244,9 +244,9 @@ SELECT emp_no FROM test_emp WHERE 1./-salary  > 1/-6E4;
 chainedNegationDenominator
```
Comment (bpintea) [inline x-pack/plugin/sql/qa/server/src/main/resources/arithmetic.sql-spec:244]: Just curious: are the changes in the spec meant to make more sense of the queries, or are there test-needs?

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java [hunk 3]
```diff
@@ -247,6 +247,41 @@ public void testComparisonAgainstColumns() {
         assertEquals("Line 1:43: Comparisons against fields are not (currently) supported; offender [int] in [>]", ex.getMessage());
     }
 
+    public void testComparisonAgainstColumnsSameSide() {
+        LogicalPlan p = plan("SELECT some.string FROM test WHERE int - float > 0");
```
Comment (bpintea) [inline x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java:251]: Just an observation: `...WHERE IIF(int - float > 0, true, false)` is currently scripted fine. Might be the case that we could suggest an alternative or check the validity.

**Review by bpintea — APPROVED** [review]
LGTM.

File: x-pack/plugin/sql/qa/server/src/main/resources/arithmetic.sql-spec (hunk 2 above)
Comment (astefan) [reply x-pack/plugin/sql/qa/server/src/main/resources/arithmetic.sql-spec:244]: Just test needs. Now, only one field is allowed inside arithmetic operations of any kind.

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java (hunk 3 above)
Comment (astefan) [reply x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java:251]: The use of multiple fields inside the same function is still permitted. The IIF query you showed will still be allowed.

Comment (bpintea) [reply x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java:251]: I think we could look at some point at this inconsistency too. Maybe later with better Painless support.

File: x-pack/plugin/sql/qa/server/src/main/resources/arithmetic.sql-spec (hunk 2 above)
Comment (bpintea) [reply x-pack/plugin/sql/qa/server/src/main/resources/arithmetic.sql-spec:244]: Right, the very scope of the PR. Thanks.

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java (hunk 3 above)
Comment (astefan) [reply x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java:251]: Yes, there's a wider applicability here. In the same ballpark as https://github.com/elastic/elasticsearch/issues/34400.

**Comment by costin** [comment]
Looks like this PR didn't make the release despite being approved and running - since it's marked as a bug, it's worth scheduling for 7.14.1

**Comment by bpintea** [comment]
> it's worth scheduling for 7.14.1

I guess porting to 7.14(.0) should still be OK, being a bug.

**Pull request closed without merging**